# Data Life Cycle (Collection to Preprocessing) in Python


Setting the Stage (The "Why")
In business, data is rarely handed to you on a silver platter. It is often "dirty" containing typos, missing entries from system glitches, or duplicate entries from multiple CRM syncs. Garbage In, Garbage Out (GIGO) is the golden rule of analytics.

Learning Objectives:
* Navigate the Python environment.

* Perform Exploratory Data Analysis (EDA) to spot errors.

* Apply cleaning techniques to make data "model-ready."

## Data Collection & Inspection

We use the pandas library to handle data. Think of a DataFrame as a programmatic Excel sheet.

### Step 1: Loading Data

In [1]:
import pandas as pd

# Loading data from a URL or local file

df = pd.read_csv("https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv")

# High-level overview
print(df.info()) # Shows data types and non-null counts


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB
None


**Show the data**

In [6]:
df

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


In [ ]:
from google.colab import sheets
sheet = sheets.InteractiveSheet(df=df)

https://docs.google.com/spreadsheets/d/1jm1ooMeZFyJFMRgWutcI5ftgw2Cns6D9l06e0CFTPzw/edit#gid=0


**Show top 5**

In [ ]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


**Shows statistical summary**

In [ ]:
print(df.describe()) # Shows statistical summary (Mean, Std, Max)

       PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200  


**What are the problems there??**

In [4]:
df['Fare'].min()

0.0



---



In [ ]:
df['Fare'].mean()
df['Fare'].describe()

,Fare
count,891.000000
mean,32.204208
std,49.693429
min,0.000000
25%,7.910400
50%,14.454200
75%,31.000000
max,512.329200


In [10]:
df.isnull().sum()

,0
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,177
SibSp,0
Parch,0
Ticket,0
Fare,0


### Step 2: The "Health Check"
Before cleaning, you must diagnose the issues.

* df.isnull().sum() : Counts how many empty cells are in each column.

* df.duplicated().sum() : Counts how many rows are exact copies.

In [12]:
# df.isnull().sum()
df.duplicated().sum()

np.int64(0)

## Data Cleaning (The "How")

**Strategy 1 - Deletion (The "Clean Slate" approach)**

Business Context: The Cabin column has 687 missing values. In the shipping industry, a missing cabin number often means the passenger had "No Cabin" (Steerage).

In [ ]:
import pandas as pd

# Loading data from a URL or local file
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pd.read_csv(url)

# High-level overview
# print(df.info()) # Shows data types and non-null counts


In [23]:
print(f"Null values in 'Age' in original df before dropping: {df['Age'].isnull().sum()}")
df = df.dropna(subset=['Age'])
print(f"Null values in 'Age' in df after dropping: {df['Age'].isnull().sum()}")

Null values in 'Age' in original df before dropping: 0
Null values in 'Age' in df after dropping: 0


In [26]:
print(f"Null values in 'Age' in original file before dropping: {df['Age'].isnull().sum()}")
age_clean= df.dropna(subset=['Age'])
print(f"Null values in 'Age' in original file after dropping: {df['Age'].isnull().sum()}")
print(f"Rows before:{len(df)},Rows after: {len(age_clean)}")


Null values in 'Age' in original file before dropping: 0
Null values in 'Age' in original file after dropping: 0
Rows before:714,Rows after: 714


In [ ]:
# Dropping rows where 'Embarked' is null
df_cleaned = df.dropna(subset=['Embarked'])

print(f"Rows before: {len(df)}, Rows after: {len(df_cleaned)}")

Rows before: 891, Rows after: 889


In [ ]:
df_cleaned['Embarked'].isnull().sum()

np.int64(0)

Clean for Age

np.int64(177)

In [ ]:
df_cleanedage = df.dropna(subset=['Age'])
print("Rows before",len(df))
print("Rows after",len(df_cleanedage))

Rows before 891
Rows after 714


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
df_cleanedage['Age'].isnull().sum()

np.int64(0)

**Strategy 2 - Constant/Categorical Imputation**

Business Context: The Cabin column has 687 missing values. In the shipping industry, a missing cabin number often means the passenger had "No Cabin" (Steerage).

In [ ]:
import pandas as pd

# Loading data from a URL or local file
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pd.read_csv(url)

# High-level overview
# print(df.info()) # Shows data types and non-null counts


In [ ]:
# Filling missing Cabin with a 'Unknown' label
df['Cabin'] = df['Cabin'].fillna('Unknown')

print("Example of filled Cabin values:")
print(df['Cabin'].head(10))

Example of filled Cabin values:
0    Unknown
1        C85
2    Unknown
3       C123
4    Unknown
5    Unknown
6        E46
7    Unknown
8    Unknown
9    Unknown
Name: Cabin, dtype: object


**Strategy 3 - Statistical Imputation (Mean vs. Median)**

Business Context: We are missing 177 ages. We can't drop them (too much data loss), so we fill them.

Mean: Good for "average" passengers.

Median: Better if there are extreme outliers (like a few 80-year-olds in a sea of 20-somethings).

In [ ]:
import pandas as pd

# Loading data from a URL or local file
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pd.read_csv(url)

# High-level overview
# print(df.info()) # Shows data types and non-null counts


In [ ]:
# Creating a copy to compare
df_mean_fill = df.copy()
df_median_fill = df.copy()

# Filling Age
df_mean_fill['Age'] = df_mean_fill['Age'].fillna(df_mean_fill['Age'].mean())
df_median_fill['Age'] = df_median_fill['Age'].fillna(df_median_fill['Age'].median())

print(f"Mean Age: {df_mean_fill['Age'].mean():.2f}")
print(f"Median Age: {df_median_fill['Age'].median():.2f}")

Mean Age: 29.36
Median Age: 28.00


**Strategy 4 - Group-Based Imputation (The "Pro" Way)**

Business Context: A 1st Class passenger is likely to be older than a 3rd Class passenger. Instead of giving everyone the global average age, let's give them the average age of their Ticket Class.

In [ ]:
import pandas as pd

# Loading data from a URL or local file
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pd.read_csv(url)

# High-level overview
# print(df.info()) # Shows data types and non-null counts


In [ ]:
# Fill missing Age based on the average age of their Pclass (1, 2, or 3)
df['Age'] = df['Age'].fillna(df.groupby('Pclass')['Age'].transform('median'))

print("Check: Are there any missing ages left?")
print(df['Age'].isnull().sum())

Check: Are there any missing ages left?
0


**Structural Cleaning**

Business data often has white spaces or inconsistent capitalization (e.g., "NY", "ny", " New York").

In [ ]:
# Standardizing text
df['Name'] = df['Name'].str.strip().str.title()
df

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S
...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,28.0,1,2,W./C. 6607,23.4500,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C


## Final Verification

In [ ]:
# Final Health Check
print("Final Missing Value Count:")
print(df.isnull().sum())

# Preview the cleaned data
df.head()

Final Missing Value Count:
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin            0
Embarked         2
dtype: int64


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,Unknown,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,Unknown,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,Unknown,S
